store/sqlstore/upgrade.go

```diff
@@ -376,6 +376,12 @@ func UpgradeDatabaseToVersion49(sqlStore SqlStore) {
 	//TODO: Uncomment the following condition when version 4.9.0 is released
 	//if shouldPerformUpgrade(sqlStore, VERSION_4_8_0, VERSION_4_9_0) {
 	sqlStore.CreateColumnIfNotExists("Teams", "LastTeamIconUpdate", "bigint", "bigint", "0")
+	if sqlStore.DriverName() == model.DATABASE_DRIVER_POSTGRES {
```

Comment:
What's happening with making this optional? Have we confirmed that we don't have any users with very large databases that will be affected by this change?

Reply:
See summary, but this was my explanation:

> No expensive migrations are performed as part of this change. If MySQL is stuck with a VARCHAR(4000), that limit remains using the old (broken) semantics of being both a count of code points and bytes. PostgreSQL is automatically transitioned from a VARCHAR(4000) to a TEXT column given that they use the same underlying types.

Going from `VARCHAR(4000)` to `TEXT` on PostgreSQL does not change the underlying data: rather, it merely removes a constraint on insertion.

See also https://www.postgresql.org/docs/current/static/datatype-character.html, discussing `CHAR`, `VARCHAR` and `TEXT`:
> There is no performance difference among these three types, apart from increased storage space when using the blank-padded type, and a few extra CPU cycles to check the length when storing into a length-constrained column. While character(n) has performance advantages in some other database systems, there is no such advantage in PostgreSQL; in fact character(n) is usually the slowest of the three because of its additional storage costs. In most situations text or character varying should be used instead.

Reply:
*However*, your comment triggered the realization that the gorp library [will still create](https://github.com/mattermost/gorp/blob/master/dialect_postgres.go#L65) /new/ Postgres databases using `VARCHAR(65535)`, so for consistency, I should probably converge on upgrading to that instead of removing the constraint altogether.

Reply:
We should still confirm what the effects of this are on a large database. The whole reason we wanted it to be optional was to prevent the upgrade from taking a long time for large databases.

We should also still look at allowing for the migration for both database types since we support both of them

Reply:
Yep, that's fair. I'm setting up a large-ish PostgreSQL database (1 million+ posts) to confirm my statements above.

Based on the gorp history, I'm very doubtful that there are many (if any) MySQL installations using `VARCHAR(4000)` -- changes there go back to 2015, IIRC. PostgreSQL is an anomaly, simply because of how gorp handles the maximum string length in terms of what columns to create. 

So in all likelihood, no MySQL migration will ever be required. But, if it's ever needed, and the feature is desired, the change would be a simple `ALTER COLUMN` statement away, easily solved via a support request for same. I'm proposing we don't write any code to do this.

Reply:
Alright, it took some more digging, but I'm now inclined to remove this migration altogether and document the required SQL to migrate to a larger column if desired. Summary below.

As of [Postgres 9.1](https://www.postgresql.org/docs/9.1/static/release-9-1.html), 
> Allow ALTER TABLE ... SET DATA TYPE to avoid table rewrites in appropriate cases (Noah Misch, Robert Haas)
> For example, converting a varchar column to text no longer requires a rewrite of the table. However, increasing the length constraint on a varchar column still requires a table rewrite.

So, /if/ a customer is running Postgres 9.1, then a `VARCHAR(4000)` to `TEXT` is safe... except that I've changed the migration to `VARCHAR(65536)` because of the way `gorp` creates the fields right now. I'm not inclined to change gorp at this time, and I can't safely migrate to `VARCHAR(65536)` without risking a table rewrite in some cases.

I did load test with 1 million posts and the migration/rewrite took a handful of seconds. Given the uncertainty here, I'd rather leave this up to the customer to transition if they require the longer post length in the first place.

Thanks for pushing back on this one @hmhealey :)